### PR TITLE
Bound initial rotary cache size

### PR DESCRIPTION
**Pull request by abhinav-bellapu**
# Description

Follow-up/prep for #1453 based on maintainer feedback there.

`HookedTransformer` rotary attention currently precomputes persistent `rotary_sin` / `rotary_cos` buffers out to full `cfg.n_ctx` for every rotary attention layer during construction. For long-context configs, that means short-context users pay the full long-context rotary-cache memory cost up front.

This PR bounds the initial rotary precompute to `min(n_ctx, 2048)` and keeps the existing `apply_rotary` extension path responsible for growing the buffers on demand when a longer position is actually used. `cfg.n_ctx` still reflects the true supported context length.

It also fixes two compatibility details needed for making lazy extension the normal path:

- `_extend_rotary_embeddings` now uses the same local-vs-global RoPE base selection as construction, so local-attention layers use `rotary_base_local` when configured.
- `_load_from_state_dict` now replaces mismatched `rotary_sin` / `rotary_cos` buffers with the current buffers, mirroring the existing mask size shim, so older checkpoints with full-size rotary buffers can still load with `strict=True`.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

## Validation

Local checks run:

```bash
uv run --no-sync pytest tests/unit/components/test_attention.py -q
make format
make check-format
uv run mypy .
make unit-test
git diff --check
make test-pr
```

Results:

- `tests/unit/components/test_attention.py`: `13 passed, 4 skipped`
- `make format`: passed
- `make check-format`: passed
- `uv run mypy .`: passed, `Success: no issues found in 299 source files`
- `make unit-test`: passed, `3220 passed, 31 skipped, 42 deselected, 9 xfailed`
- `git diff --check`: passed
- `make test-pr` progressed through:
  - unit: passed, `3220 passed, 31 skipped, 42 deselected, 9 xfailed`
  - docstring: passed, `18 passed, 25 skipped`
  - acceptance: passed, `137 passed, 83 skipped, 11 deselected, 1 rerun`
  - integration: blocked locally before test execution by a macOS temporary-venv import hang in the notebook pytest plugin path (`nbval` / `jupyter_client` / `zmq`). CI should run this in a clean environment.

### CODE REVIEW

**Comment by jlarson4** [comment]
Great work! Approved and merging.

One additional note that's outside the scope of what I recommended originally:

`_extend_rotary_embeddings` recomputes the entire buffer on each growth, and the trigger extends to exactly `max_pos_needed`. For a long-prompt prefill that's one extension, but for token-by-token generation beyond 2048 (KV cache, `x_pos=1`, `max_pos_needed` +1/step), every step recomputes the whole growing buffer which means ~`O(n^2)` rotary recompute over a long generation. It's pre-existing behavior, but this PR makes lazy extension the normal path for exactly the long-context models that would hit it. A cheap mitigation: grow with headroom (example: `min(n_ctx, max(max_pos_needed, 2 * current_size))`) so extension fires `O(log n)` times instead of every step. 

Feel free to do another follow up PR if you're interested!
